File: flask_app/app/dao/user_tier_metrics_dao.py
from datetime import datetime
from app.models.user_tier_metrics import UserTierMetrics
from app.models import db


class UserTierMetricsDAO:

    @staticmethod
    def get_for_user_month(user_id, month_year):
        return UserTierMetrics.query.filter_by(
            user_id=user_id, month_year=month_year
        ).first()

    @staticmethod
    def create_or_update(data):
        user_id = data["user_id"]
        month_year = data.get('month_year') or datetime.utcnow().strftime("%Y-%m")


        utm = UserTierMetricsDAO.get_for_user_month(user_id, month_year)

        if utm:
            utm.order_count += data.get("order_count", 0)
            utm.total_spent += data.get("total_spent", 0)
        else:
            utm = UserTierMetrics(
                user_id=user_id,
                month_year=month_year,
                order_count=data.get("order_count", 0),
                total_spent=data.get("total_spent", 0.0),
            )
            db.session.add(utm)
        db.session.commit()
        return utm

    @staticmethod
    def get_all_metrics():
        return UserTierMetrics.query.order_by(UserTierMetrics.month_year.desc()).all()

    @staticmethod
    def get_top_spenders(month_year, top_n=10):
        return (
            UserTierMetrics.query.filter_by(month_year=month_year)
            .order_by(UserTierMetrics.total_spent.desc())
            .limit(top_n)
            .all()
        )

File: springboot_api/src/main/java/com/example/membership_api/dao/UserTierMetricsDAO.java
package com.example.membership_api.dao;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.List;

import org.springframework.stereotype.Repository;

import com.example.membership_api.model.UserTierMetrics;

import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.transaction.Transactional;

@Repository
@Transactional
public class UserTierMetricsDAO {

    @PersistenceContext
    private EntityManager em;

    public UserTierMetrics getForUserMonth(String userId, String monthYear) {
        List<UserTierMetrics> results = em.createQuery(
                "FROM UserTierMetrics u WHERE u.userId = :userId AND u.monthYear = :monthYear", UserTierMetrics.class)
                .setParameter("userId", userId)
                .setParameter("monthYear", monthYear)
                .getResultList();

        return results.isEmpty() ? null : results.get(0);
    }

    public UserTierMetrics createOrUpdate(String userId, int orderCount, double totalSpent) {
        String monthYear = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM"));

        UserTierMetrics utm = getForUserMonth(userId, monthYear);

        if (utm != null) {
            utm.setOrderCount(utm.getOrderCount() + orderCount);
            utm.setTotalSpent(utm.getTotalSpent() + totalSpent);
        } else {
            utm = new UserTierMetrics();
            utm.setUserId(userId);
            utm.setMonthYear(monthYear);
            utm.setOrderCount(orderCount);
            utm.setTotalSpent(totalSpent);
            em.persist(utm);
        }

        return utm;
    }

    public List<UserTierMetrics> getAllMetrics() {
        return em.createQuery("FROM UserTierMetrics u ORDER BY u.monthYear DESC", UserTierMetrics.class)
                 .getResultList();
    }

    public List<UserTierMetrics> getTopSpenders(String monthYear, int topN) {
        return em.createQuery("FROM UserTierMetrics u WHERE u.monthYear = :monthYear ORDER BY u.totalSpent DESC", UserTierMetrics.class)
                 .setParameter("monthYear", monthYear)
                 .setMaxResults(topN)
                 .getResultList();
    }
}
